File: kits/js/bench/carriage.test.js
// The common carriage: the hint as the whole address, one POST, bytes in and
// bytes out, and nothing else HTTP offers. Two wardens stand up on ephemeral
// loopback ports and drive a real grant, ask, answer and news across the wire.
import test from 'node:test';
import assert from 'node:assert/strict';
import { request } from 'node:http';
import {
  Warden,
  commitment,
  encode,
  news,
  post,
  readAnswer,
  readField,
  signingPair,
} from '../src/index.js';
import { serve } from '../src/door.js';
import { host as stand } from '../src/host.js';

const hex = (bytes) => Buffer.from(bytes).toString('hex');
const fixed = (fill) => new Uint8Array(32).fill(fill);
const utf8 = new TextEncoder();
const still = () => 1_000;
const RANDOM = fixed(200);
const TEXT = { base: 'text' };

const LIST = `ToDo
  add(title text) bool
  count() int
`;

function todo() {
  return {
    lines: [],
    add(title) {
      this.lines.push(title);
      return true;
    },
    count() {
      return BigInt(this.lines.length);
    },
  };
}

// Two grounds, each with its own door on its own ephemeral loopback port. Two
// wardens are strangers: same machine, same seals, same judgment.
async function grounds() {
  // The clock and the randomness are handed to the warden, never reached for:
  // the door behind a road takes them from the warden it serves.
  let grain = 100;
  const random = () => fixed((grain += 1) % 251);
  const host = await Warden.open({
    nameSeed: fixed(1),
    padlockSeed: fixed(2),
    heirSeed: fixed(3),
    clock: still,
    random,
  });
  const guest = await Warden.open({
    nameSeed: fixed(10),
    padlockSeed: fixed(11),
    heirSeed: fixed(12),
    clock: still,
    random,
  });
  const object = todo();
  const { being } = await host.hold(object, {
    seed: fixed(5),
    heirSeed: fixed(6),
    blueprint: LIST,
  });

  // The door tells the warden where it ended up: neither address exists until
  // the socket is bound, and everything minted after this carries them.
  const there = await serve(host);
  const here = await serve(guest);
  return {
    host,
    guest,
    object,
    being,
    there,
    here,
    close: () => Promise.all([there.close(), here.close()]),
  };
}

async function readBack(guest, host, envelope, field) {
  if (envelope === null) return null;
  const answer = await readAnswer({
    envelope,
    padlockSecret: guest.padlock.secret,
    wardenPk: host.name.pk,
  });
  if (!answer) return null;
  return field ? readField(field, answer.data) : answer;
}

test('a grant, an ask and an answer cross the wire as bytes in and bytes out', async (t) => {
  const world = await grounds();
  t.after(world.close);
  const { host, guest, object, being } = world;

  // The invitation changes hands as data, carrying the hints — the only thing
  // that can, because delivery cannot begin from nothing.
  const invitation = await host.grant(being, { voiceSeed: fixed(20), heirSeed: fixed(21) });
  assert.deepEqual(invitation.hints, [world.there.hint]);
  const row = guest.remember(invitation);

  // The holder's first act: rotate, ask nothing, and read back what it stands
  // at — over the wire, at the hint and nowhere else.
  const next = await signingPair(fixed(22));
  const estate = await readBack(
    guest,
    host,
    await post(
      row.hints[0],
      await guest.ask(row, {
        seq: 1n,
        commitment: await commitment(host.name.pk, next.pk),
        random: RANDOM,
      }),
    ),
    'describe',
  );
  row.voice = { pk: row.heir.pk, secret: row.heir.secret };
  assert.equal(estate.classes.length, 2);
  assert.ok(
    estate.classes.some((one) => one.beings.some((held) => hex(held.being) === hex(being))),
  );

  // And an ordinary ask on the being, answered across the same one POST.
  const answer = await readBack(
    guest,
    host,
    await post(
      world.there.hint,
      await guest.ask(row, {
        seq: 2n,
        being,
        method: { name: 'add', args: encode(TEXT, 'milk') },
        random: RANDOM,
      }),
    ),
  );
  assert.equal(answer.seq, 2n);
  assert.deepEqual(object.lines, ['milk']);
});

test('news crosses the wire in the other direction, judged by the same steps', async (t) => {
  const world = await grounds();
  t.after(world.close);
  const { host, guest } = world;

  // The guest holds a relation with the host's house, so the host's own name
  // is placed in the guest's outbound record: what arrives is news.
  const { being } = await host.hold(todo(), {
    seed: fixed(7),
    heirSeed: fixed(8),
    blueprint: LIST,
  });
  const invitation = await host.grant(being, { voiceSeed: fixed(30), heirSeed: fixed(31) });
  guest.remember(invitation);

  const lock = fixed(70);
  const answer = await post(
    world.here.hint,
    await news(host, {
      // A warden must be able to reach a peer it is not answering. Here it
      // holds the peer's name as well as its lock, so the recipient is the
      // name; a peer known only by its padlock is named by that instead.
      peer: { name: guest.name.pk, padlock: guest.padlock.pk },
      voice: host.name,
      word: {
        being: null,
        successor: null,
        commitment: null,
        name: null,
        padlock: lock,
        hints: ['https://moved.example'],
      },
      seq: 1n,
      random: RANDOM,
    }),
  );
  // `tell` answers nothing, so what comes back is an answer with no data.
  const read = await readAnswer({
    envelope: answer,
    padlockSecret: host.padlock.secret,
    wardenPk: guest.name.pk,
  });
  assert.equal(read.data, null);
  assert.equal(hex((await guest.outboundFor(host.name.pk)).padlock), hex(lock));
  assert.deepEqual((await guest.outboundFor(host.name.pk)).hints, ['https://moved.example']);
});

test('silence is an empty body, and no status code carries meaning', async (t) => {
  const world = await grounds();
  t.after(world.close);
  const { host, guest, being } = world;

  // A voice in neither record, at a being it does not reach.
  const stranger = await signingPair(fixed(50));
  const envelope = await guest.carry({
    recipient: host.name.pk,
    padlock: host.padlock.pk,
    voicePk: stranger.pk,
    voiceSecret: stranger.secret,
    seq: 1n,
    allowance: { time: 5_000n, hops: 4n },
    being,
    method: { name: 'count', args: new Uint8Array(0) },
    random: RANDOM,
  });
  assert.equal(await post(world.there.hint, envelope), null);

  // The door answers the same way to bytes that are no box at all — a refusal
  // and an answer leave by the same door, and neither narrates.
  assert.equal(await post(world.there.hint, new Uint8Array(0)), null);
  assert.equal(await post(world.there.hint, utf8.encode('not a box')), null);

  // And nothing else of HTTP is read or written: the status is 200 whether the
  // door spoke or was silent, and the body is the whole of the message.
  const status = await statusOf(world.there.hint, new Uint8Array(0));
  assert.equal(status.code, 200);
  assert.equal(status.length, 0);
  const spoke = await statusOf(
    world.there.hint,
    await guest.carry({
      recipient: host.name.pk,
      padlock: host.padlock.pk,
      voicePk: stranger.pk,
      voiceSecret: stranger.secret,
      seq: 1n,
      allowance: { time: 5_000n, hops: 4n },
      being: null,
      method: null,
      random: RANDOM,
    }),
  );
  assert.equal(spoke.code, 200);
  assert.ok(spoke.length > 0);
});

test('the hint is posted to exactly as given, with no path and no query', async (t) => {
  const world = await grounds();
  t.after(world.close);
  const seen = [];
  world.there.server.on('request', (incoming) => seen.push(incoming.url));
  const stranger = await signingPair(fixed(50));
  await post(
    world.there.hint,
    await world.guest.carry({
      recipient: world.host.name.pk,
      padlock: world.host.padlock.pk,
      voicePk: stranger.pk,
      voiceSecret: stranger.secret,
      seq: 1n,
      allowance: { time: 5_000n, hops: 4n },
      being: null,
      method: null,
      random: RANDOM,
    }),
  );
  // No path was appended and no query added: the URL the hint named is the
  // whole of the address.
  assert.deepEqual(seen, ['/']);
});

test('a road that never carried the bytes is weather, not silence', async () => {
  // Silence has a wire form on this carriage: an empty body. A connection
  // refused and a name that does not resolve said neither of those, so the kit
  // reports the road's fault rather than inventing an empty body.
  const envelope = new Uint8Array(48);
  await assert.rejects(() => post('http://127.0.0.1:1/', envelope));
  await assert.rejects(() => post('http://a-door-that-is-not.invalid/', envelope));
});

test('a caller tries the hints it holds, because none is authoritative', async (t) => {
  // A hint is a guess about the weather, and delivery is what tries them: the
  // dead road is tried and the live one carries. Nothing is proved by the
  // arrival — everything is proved by the seal.
  const alice = await stand({
    seeds: { name: fixed(60), padlock: fixed(61), heir: fixed(62) },
    clock: still,
    random: () => crypto.getRandomValues(new Uint8Array(32)),
    roads: ['http'],
  });
  const bob = await stand({
    seeds: { name: fixed(63), padlock: fixed(64), heir: fixed(65) },
    clock: still,
    random: () => crypto.getRandomValues(new Uint8Array(32)),
  });
  t.after(() => Promise.all([alice.close(), bob.close()]));

  const object = todo();
  await alice.warden.hold(object, { blueprint: LIST });
  const invitation = await object.quo.grant(object);
  // A road that is nothing but weather, offered first.
  invitation.hints.unshift('http://127.0.0.1:1/');
  const [handle] = await bob.warden.accept(invitation, { label: 'todo' });
  assert.equal(await handle.add('milk'), true);
  assert.deepEqual(object.lines, ['milk']);
});

test('a door that learns its address at serve time mints invitations that reach it', async (t) => {
  // Opened knowing nothing about where it will stand: an ephemeral port has no
  // address until the socket is bound, so a warden that fixed its roads at
  // birth could only hand out hints that reach nobody.
  let grain = 60;
  const house = await Warden.open({
    nameSeed: fixed(30),
    padlockSeed: fixed(31),
    heirSeed: fixed(32),
    clock: still,
    random: () => fixed((grain += 1) % 251),
  });
  assert.deepEqual(house.hints, []);
  const { being } = await house.hold(todo(), {
    seed: fixed(33),
    heirSeed: fixed(34),
    blueprint: LIST,
  });
  const caller = await Warden.open({
    nameSeed: fixed(40),
    padlockSeed: fixed(41),
    heirSeed: fixed(42),
  });

  const door = await serve(house);
  t.after(door.close);
  assert.deepEqual(house.hints, [door.hint]);

  const invitation = await house.grant(being, { voiceSeed: fixed(43), heirSeed: fixed(44) });
  assert.deepEqual(invitation.hints, [door.hint]);

  // The proof is not that the strings match but that the road carries: the
  // holder posts down the hints it was given and a door answers.
  const row = caller.remember(invitation);
  const next = await signingPair(fixed(45));
  const estate = await readBack(
    caller,
    house,
    await post(
      row.hints[0],
      await caller.ask(row, {
        seq: 1n,
        commitment: await commitment(house.name.pk, next.pk),
        random: RANDOM,
      }),
    ),
    'describe',
  );
  const stands = estate.classes.flatMap((one) => one.beings.map((each) => hex(each.being)));
  assert.ok(stands.includes(hex(being)));

  // A second road is added, never swapped: a warden offers as many as it has.
  house.publish('https://house.example');
  const card = house.card();
  assert.deepEqual(card.hints, [door.hint, 'https://house.example']);

  // And a road that stops carrying stops being minted: the second door is the
  // whole of a warden moving house, which is open the new road, close the old.
  const moved = await serve(house);
  t.after(moved.close);
  assert.deepEqual(house.hints, [door.hint, 'https://house.example', moved.hint]);
  await moved.close();
  assert.deepEqual(house.card().hints, [door.hint, 'https://house.example']);
});

function statusOf(hint, body) {
  return new Promise((resolve, reject) => {
    const call = request(hint, { method: 'POST' }, (incoming) => {
      const chunks = [];
      incoming.on('data', (chunk) => chunks.push(chunk));
      incoming.on('end', () =>
        resolve({ code: incoming.statusCode, length: Buffer.concat(chunks).length }),
      );
    });
    call.on('error', reject);
    call.end(Buffer.from(body));
  });
}

test('the published limit and the enforced one are two separate acts, and nothing joins them', async (t) => {
  // The warden's part is only to publish the number; enforcing it is the road's
  // and the operator holds the two together by hand. Nothing derives one from
  // the other, so a house can publish a generous number and stand behind a mean
  // door — and every honest caller doing the arithmetic the published number
  // invites is refused in silence, which is what makes it undiagnosable.
  const published = 65_536n;
  const enforced = 4_096n;

  let grain = 150;
  const house = await Warden.open({
    nameSeed: fixed(70),
    padlockSeed: fixed(71),
    heirSeed: fixed(72),
    limit: published,
    clock: still,
    random: () => fixed((grain += 1) % 251),
  });
  const { being } = await house.hold(todo(), {
    seed: fixed(73),
    heirSeed: fixed(74),
    blueprint: LIST,
  });
  const caller = await Warden.open({
    nameSeed: fixed(75),
    padlockSeed: fixed(76),
    heirSeed: fixed(77),
  });

  const door = await serve(house, { limit: enforced });
  t.after(door.close);

  const invitation = await house.grant(being, { voiceSeed: fixed(78), heirSeed: fixed(79) });
  const row = caller.remember(invitation);
  const next = await signingPair(fixed(80));
  await post(
    door.hint,
    await caller.ask(row, {
      seq: 1n,
      commitment: await commitment(house.name.pk, next.pk),
      random: RANDOM,
    }),
  );
  row.voice = { pk: row.heir.pk, secret: row.heir.secret };
  row.heir = next;

  // What the house says about itself, asked over the wire like anything else.
  const asked = await readBack(
    caller,
    house,
    await post(
      door.hint,
      await caller.ask(row, {
        seq: 2n,
        being: house.name.pk,
        method: { name: 'limit', args: new Uint8Array(0) },
        random: RANDOM,
      }),
    ),
    'limit',
  );
  assert.equal(asked, published, 'the warden publishes the number it was opened with');

  // A caller that believes it and sends something comfortably inside it meets
  // silence, because the road in front of the door was held to a smaller
  // number and the road is what counts the bytes.
  const big = encode(TEXT, 'x'.repeat(Number(enforced)));
  const over = await caller.ask(row, {
    seq: 3n,
    being,
    method: { name: 'add', args: big },
    random: RANDOM,
  });
  assert.ok(
    BigInt(over.length) > enforced && BigInt(over.length) < published,
    'the probe has to sit between the two numbers or it proves nothing',
  );
  assert.equal(await post(door.hint, over), null);

  // And it is the road rather than the judgment: the same envelope handed
  // straight to the warden is answered, so nothing about the message was wrong.
  assert.notEqual(await house.judge(over, { clock: still, random: RANDOM }), null);
});

File: kits/zig/test/carriage_test.zig
//! The common carriage, asserted from Article III alone — there is no corpus
//! for it. Every test is named for the clause it pins, and a refusal is
//! asserted as strictly as an acceptance.
//!
//! The road is exercised over a real HTTP door bound on an ephemeral
//! loopback port and torn down with the suite. It is HTTP rather than HTTPS
//! because TLS is redundant crypto Quo relies on for no guarantee — the
//! carriage's promises are about the POST, the body and the silence, and
//! none of them is a TLS promise.

const std = @import("std");
const carriage = @import("carriage");
const src_dir = @import("sources").src_dir;

const net = std.Io.net;

/// A door that records what arrived and answers whatever it was told to.
const Recorder = struct {
    gpa: std.mem.Allocator,
    /// What the door was handed, if it was handed anything.
    heard: ?[]u8 = null,
    knocks: usize = 0,
    /// The bytes to answer with, or `null` for silence.
    answer: ?[]const u8 = null,

    fn knock(context: *anyopaque, gpa: std.mem.Allocator, sealed: []const u8) std.mem.Allocator.Error!carriage.Answer {
        const self: *Recorder = @ptrCast(@alignCast(context));
        self.knocks += 1;
        if (self.heard) |old| self.gpa.free(old);
        self.heard = try self.gpa.dupe(u8, sealed);
        const reply = self.answer orelse return null;
        return try gpa.dupe(u8, reply);
    }

    fn door(self: *Recorder) carriage.Door {
        return .{ .context = self, .knock = knock };
    }

    fn deinit(self: *Recorder) void {
        if (self.heard) |h| self.gpa.free(h);
    }
};

const Ground = struct {
    server: net.Server,
    io: std.Io,
    gpa: std.mem.Allocator,
    door: carriage.Door,
    /// How many requests to answer before the ground goes away.
    rounds: usize = 1,
    failed: bool = false,

    fn run(self: *Ground) void {
        var left = self.rounds;
        while (left > 0) : (left -= 1) {
            const stream = self.server.accept(self.io) catch {
                self.failed = true;
                return;
            };
            defer stream.close(self.io);
            carriage.serveOnce(self.gpa, self.io, stream, 1 << 20, self.door) catch {
                self.failed = true;
                return;
            };
        }
    }

    fn hint(self: *Ground, gpa: std.mem.Allocator) ![]u8 {
        return std.fmt.allocPrint(gpa, "http://127.0.0.1:{d}/", .{self.server.socket.address.getPort()});
    }
};

fn loopback(io: std.Io) !net.Server {
    const address: net.IpAddress = .{ .ip4 = .loopback(0) };
    return carriage.listen(io, &address);
}

test "III: one POST, bytes in and bytes out, and the response body is the sealed answer" {
    const gpa = std.testing.allocator;
    var threaded: std.Io.Threaded = .init(gpa, .{});
    defer threaded.deinit();
    const io = threaded.io();

    var recorder: Recorder = .{ .gpa = gpa, .answer = "the sealed answer" };
    defer recorder.deinit();

    var ground: Ground = .{
        .server = loopback(io) catch return error.SkipZigTest,
        .io = io,
        .gpa = gpa,
        .door = recorder.door(),
    };
    defer ground.server.deinit(io);

    const hint = try ground.hint(gpa);
    defer gpa.free(hint);

    const thread = try std.Thread.spawn(.{}, Ground.run, .{&ground});
    const answer = try carriage.post(gpa, io, hint, "the sealed ask", 1 << 20);
    thread.join();

    try std.testing.expect(!ground.failed);
    try std.testing.expectEqual(@as(usize, 1), recorder.knocks);
    try std.testing.expectEqualStrings("the sealed ask", recorder.heard.?);
    try std.testing.expect(answer != null);
    defer gpa.free(answer.?);
    try std.testing.expectEqualStrings("the sealed answer", answer.?);
}

test "III: an empty body is silence's wire form" {
    const gpa = std.testing.allocator;
    var threaded: std.Io.Threaded = .init(gpa, .{});
    defer threaded.deinit();
    const io = threaded.io();

    // The door answers nothing, which is what a failed judgment looks like at
    // every one of its eight steps.
    var recorder: Recorder = .{ .gpa = gpa, .answer = null };
    defer recorder.deinit();

    var ground: Ground = .{
        .server = loopback(io) catch return error.SkipZigTest,
        .io = io,
        .gpa = gpa,
        .door = recorder.door(),
    };
    defer ground.server.deinit(io);

    const hint = try ground.hint(gpa);
    defer gpa.free(hint);

    const thread = try std.Thread.spawn(.{}, Ground.run, .{&ground});
    const answer = try carriage.post(gpa, io, hint, "the sealed ask", 1 << 20);
    thread.join();

    try std.testing.expect(!ground.failed);
    try std.testing.expectEqual(@as(usize, 1), recorder.knocks);
    // The door was reached and said nothing, and the caller cannot tell that
    // apart from any other failure — which is the point.
    try std.testing.expectEqual(@as(carriage.Answer, null), answer);
}

test "III: anything that is not a POST of a sealed body meets the same silence" {
    const gpa = std.testing.allocator;
    var threaded: std.Io.Threaded = .init(gpa, .{});
    defer threaded.deinit();
    const io = threaded.io();

    var recorder: Recorder = .{ .gpa = gpa, .answer = "would have answered" };
    defer recorder.deinit();

    var ground: Ground = .{
        .server = loopback(io) catch return error.SkipZigTest,
        .io = io,
        .gpa = gpa,
        .door = recorder.door(),
        .rounds = 3,
    };
    defer ground.server.deinit(io);

    const port = ground.server.socket.address.getPort();
    const url = try std.fmt.allocPrint(gpa, "http://127.0.0.1:{d}/", .{port});
    defer gpa.free(url);

    const thread = try std.Thread.spawn(.{}, Ground.run, .{&ground});

    var client: std.http.Client = .{ .allocator = gpa, .io = io };
    defer client.deinit();

    // A verb with no body, and two carrying the very bytes a POST would have
    // carried — the verb is the whole of what makes them silence.
    const wrong = [_]struct { verb: std.http.Method, payload: ?[]const u8 }{
        .{ .verb = .GET, .payload = null },
        .{ .verb = .PUT, .payload = "the sealed ask" },
        .{ .verb = .PATCH, .payload = "the sealed ask" },
    };
    for (wrong) |attempt| {
        var body: std.Io.Writer.Allocating = .init(gpa);
        defer body.deinit();
        _ = try client.fetch(.{
            .location = .{ .url = url },
            .method = attempt.verb,
            .payload = attempt.payload,
            .keep_alive = false,
            .redirect_behavior = .unhandled,
            .response_writer = &body.writer,
        });
        // Silence, in the only form this road has for it.
        try std.testing.expectEqual(@as(usize, 0), body.written().len);
    }
    thread.join();

    try std.testing.expect(!ground.failed);
    // And the door was never troubled: a verb that is not POST carries no
    // unsealable bytes, so nothing above the carriage ever learned of it.
    try std.testing.expectEqual(@as(usize, 0), recorder.knocks);
}

test "III: no status code carries meaning, so an answer under one is still the answer" {
    const gpa = std.testing.allocator;
    var threaded: std.Io.Threaded = .init(gpa, .{});
    defer threaded.deinit();
    const io = threaded.io();

    var ground = try oddStatusGround(io);
    defer ground.server.deinit(io);

    const hint = try std.fmt.allocPrint(gpa, "http://127.0.0.1:{d}/", .{ground.server.socket.address.getPort()});
    defer gpa.free(hint);

    const thread = try std.Thread.spawn(.{}, OddStatus.run, .{&ground});
    const answer = try carriage.post(gpa, io, hint, "the sealed ask", 1 << 20);
    thread.join();

    try std.testing.expect(answer != null);
    defer gpa.free(answer.?);
    // A 500 with bytes in it is bytes; the caller never looked at the number.
    try std.testing.expectEqualStrings("sealed under a five hundred", answer.?);
}

/// A door that answers with a status nobody should read. It is written here
/// rather than in `carriage.zig` because the kit's own door has one status
/// and needs no other.
const OddStatus = struct {
    server: net.Server,
    io: std.Io,

    fn run(self: *OddStatus) void {
        const stream = self.server.accept(self.io) catch return;
        defer stream.close(self.io);
        var in: [4096]u8 = undefined;
        var out: [4096]u8 = undefined;
        var reader = stream.reader(self.io, &in);
        var writer = stream.writer(self.io, &out);
        var server = std.http.Server.init(&reader.interface, &writer.interface);
        var request = server.receiveHead() catch return;
        var body_buf: [4096]u8 = undefined;
        const body_reader = request.readerExpectNone(&body_buf);
        _ = body_reader.discardRemaining() catch {};
        request.respond("sealed under a five hundred", .{
            .status = .internal_server_error,
            .keep_alive = false,
        }) catch {};
    }
};

fn oddStatusGround(io: std.Io) !OddStatus {
    return .{ .server = loopback(io) catch return error.SkipZigTest, .io = io };
}

test "III: many wardens may stand behind one hint, and the carriage never reads one" {
    // Quo never parses a hint: it is an opaque string the carriage
    // understands. The carriage's only claim is that it posts to the string
    // it was given, so a string that is not a URL is the carriage's own
    // failure and never a protocol judgment.
    const gpa = std.testing.allocator;
    var threaded: std.Io.Threaded = .init(gpa, .{});
    defer threaded.deinit();
    const io = threaded.io();

    try std.testing.expectError(
        error.InvalidFormat,
        carriage.post(gpa, io, "not a url at all", "bytes", 1 << 20),
    );
    try std.testing.expectError(
        error.UnsupportedUriScheme,
        carriage.post(gpa, io, "tcp://127.0.0.1:1", "bytes", 1 << 20),
    );
}

fn readSource(gpa: std.mem.Allocator, name: []const u8) ![]u8 {
    const path = try std.fs.path.join(gpa, &.{ src_dir, name });
    defer gpa.free(path);
    var threaded: std.Io.Threaded = .init(gpa, .{});
    defer threaded.deinit();
    return std.Io.Dir.cwd().readFileAlloc(threaded.io(), path, gpa, .limited(1 << 20));
}

test "nothing above the roads reaches a host: the carriage is where std.http lives" {
    const gpa = std.testing.allocator;

    const above = [_][]const u8{
        "notation.zig",
        "arithmetic.zig",
        "wire.zig",
        "envelope.zig",
        "warden.zig",
        "quo.zig",
    };
    for (above) |name| {
        const text = try readSource(gpa, name);
        defer gpa.free(text);
        try std.testing.expect(std.mem.indexOf(u8, text, "std.http") == null);
        try std.testing.expect(std.mem.indexOf(u8, text, "Io.net") == null);
    }

    const road = try readSource(gpa, "carriage.zig");
    defer gpa.free(road);
    try std.testing.expect(std.mem.indexOf(u8, road, "std.http") != null);
}
